openvpn/middlewares/server/auth/middleware.go

```diff
@@ -0,0 +1,153 @@
+package auth
+
+import (
+	"fmt"
+	log "github.com/cihub/seelog"
+	"github.com/mysterium/node/openvpn"
+	"github.com/mysterium/node/openvpn/middlewares"
+	"net"
+	"regexp"
+	"strconv"
+)
+
+type middleware struct {
+	authenticator Authenticator
+	connection    net.Conn
+	username      string
+	password      string
+	clientId      int
+	keyId         int
+	state         middlewares.State
+}
+
+// NewMiddleware creates server user_auth challenge authentication middleware
+func NewMiddleware(authenticator Authenticator) openvpn.ManagementMiddleware {
+	return &middleware{
+		authenticator: authenticator,
+		connection:    nil,
+	}
+}
+
+func (m *middleware) Start(connection net.Conn) error {
+	m.connection = connection
+
+	_, err := m.connection.Write([]byte("state on all\n"))
+	return err
+}
+
+func (m *middleware) Stop() error {
+	_, err := m.connection.Write([]byte("state off all\n"))
+	return err
+}
+
+func (m *middleware) State() middlewares.State {
```

Comment:
How can server have one single state? It has state for each client, is not it?

Reply:
No, management interface processes all incoming request serially. Looks like auth calls are atomic.